DM-24734: Rework QuantumGraph generation to avoid O(N^2) scaling.

TallJimbo opened the pull request.

natelust commented on python/lsst/pipe/base/graphBuilder.py:
```diff
@@ -564,108 +586,80 @@ def fillDatasetRefs(self, registry, collections, run, *, skipExisting=True):
+            _LOG.debug("Resolving %d datasets for input dataset %s.", len(refs), datasetType.name)
+            for dataId in refs:
+                refs[dataId] = registry.findDataset(datasetType, dataId=dataId, collections=collections)
+            assert not any(ref is None for ref in refs.values())
```
bare asserts can be hard for debug. Its up to you to decide if it is likely enough to happen to make a richer message worth it.

TallJimbo replied:
Yeah, this one is probably worth changing to a `RuntimeError` raise, as it's guarding against both daf_butler logic (which is probably too "far away" for this assert), _and_ that these datasets haven't been deleted by some other Registry client since the previous query.

natelust commented on python/lsst/pipe/base/graphBuilder.py:
```diff
@@ -564,108 +586,80 @@ def fillDatasetRefs(self, registry, collections, run, *, skipExisting=True):
+                    elif not skipExisting:
+                        raise OutputExistsError(f"Output dataset {datasetType.name} already exists in "
+                                                f"output RUN collection '{run}' with data ID {dataId}.")
+                    refs[dataId] = ref
```
it seems weird to pull out the ref and then reassign it in the case it is not resolved. I think these branches can be re worked a bit.

timj commented on python/lsst/pipe/base/pipelineIR.py:
```diff
@@ -440,7 +440,7 @@ def to_primitives(self):
         if self.instrument is not None:
             accumulate['instrument'] = self.instrument
-        accumulate['tasks'] = {l: t.to_primitives() for l, t in self.tasks.items()}
+        accumulate['tasks'] = {m: t.to_primitives() for m, t in self.tasks.items()}
```
`l` is frowned upon as a variable because some fonts confuse it with `1`

TallJimbo commented on python/lsst/pipe/base/graphBuilder.py:
```diff
@@ -191,22 +126,143 @@ def dimensions(self) -> DimensionGraph:
+        dataIds : `Iterable` [ `DataCoordinate` ]
+            Data IDs to match.
+
+        Yields
```
Agreed, in the sense that this is what the Python community expects, but I'm curious as to whether you know if this actually matters.  AFAICT, whether an iterator-returning function is implemented as a generator or not is completely transparent to the caller, so distinguishing between them is always _just_ exposing an implementation detail.  Is that wrong in some edge cases (or maybe coroutines)?

natelust replied:
A few thoughts:
* pedantically the function does not yield anything it returns something that yields something. Its like the difference between iterable and iterator
* If someone really wanted to know the call stack (or something like that), this will never be re-entered, the thing you are returning will be
* It does make a difference when coroutines are involved (in so far as you loose the richer type information provided to any checkers)

I think in day to day practice you can use this as is, but your terminology does not really match what is happening, and if it was used like this in other places (co routines) it would be wrong. Code in a generator function is not executed until the first use (everything up to the first yield, which is why it must be "primed" with a next, but in this case there is code execution each time before the generator is returned. If someone really cared about that for some reason then the info you wrote would lead them astray. In practice it probably doesnt matter much.

TallJimbo commented on python/lsst/pipe/base/graphBuilder.py:
```diff
@@ -191,22 +126,143 @@ def dimensions(self) -> DimensionGraph:
+    dataId : `DataCoordinate`
+        Data ID for this quantum.
+    """
+    def __init__(self, task: _TaskScaffolding, dataId: DataCoordinate):
```
I think taking that approach (along with the changes in your next few comments) also requires adding `field(init=False)` or defaults (and hence dropping `__slots__`) for `inputs`, `outputs`, and `prerequisites`, and that `task` _shouldn't_ actually be `InitVar` in that scenario since it's also an attribute.  Overall it seems simpler to just define my own `__init__` than go to so much work to make the `dataclass-injected` one usable for me.

Maybe the better question is whether I should drop `@dataclass` - I'm already overriding `__repr__` and `__init__`, so all I'm getting from that is equality comparison, and I don't actually want that, I suppose.  Unless I hear otherwise I'll just drop the decorator.

TallJimbo commented on python/lsst/pipe/base/graphBuilder.py:
```diff
@@ -226,44 +282,23 @@ class _TaskScaffolding:
-                                    f"the pipeline dimensions {parent.dimensions}.")
-
-        # Initialize _DatasetScaffoldingDicts as subsets of the one or two
+        assert self.dimensions.issubset(parent.dimensions)
```
It's not dependent on user inputs - it's a check that logic in `_PipelineScaffolding.__init__` that is supposed to guarantee this does in fact do so.

TallJimbo commented on python/lsst/pipe/base/graphBuilder.py:
```diff
@@ -279,57 +314,37 @@ def __repr__(self):
-    """All quanta for this task in the graph (`list` of `Quantum`).
-
-    Populated after construction by `_PipelineScaffolding.fillQuanta`.
+    quanta: Dict[DataCoordinate, _QuantumScaffolding]
```
Are you bothered by the fact that this is a mapping and not some other kind of collection, or the fact that the value type is not `Quantum`?  I'm more sympathetic to the latter, but it's the pattern I've followed elsewhere in this file (e.g. "tasks" is a collection of `_TaskScaffolding`), and I think alternative is a lot of "Scaffolding" suffixes on variables.  That feels like visual noise to me, because I'm already familiar with this code, but it probably wouldn't be to someone reading it for the first time, and I guess the question is who to optimize readability for.  This code isn't going to be understandable at a glance regardless, so I lean towards making it more readable for those who have already put some time into studying it.

natelust replied:
the latter, as it is just hard to understand what to expect from reading variable names places. It is probably not worth a large overhaul around, but something to keep in mind.

TallJimbo commented on python/lsst/pipe/base/graphBuilder.py:
```diff
@@ -498,44 +503,61 @@ def fillDataIds(self, registry, collections, userQuery):
-        _LOG.debug("Finished processing results from data ID query.")
-
-    def fillDatasetRefs(self, registry, collections, run, *, skipExisting=True):
+        # Iterate over query results and populate the data IDs in TODO
```
Ah, yup, this was a comment I knew I was invalidating and intended to go back and rewrite, but apparently never did.  Will fix.

TallJimbo commented on python/lsst/pipe/base/graphBuilder.py:
```diff
@@ -498,44 +503,61 @@ def fillDataIds(self, registry, collections, userQuery):
+            for datasetType, refs in itertools.chain(self.inputs.items(), self.intermediates.items(),
+                                                     self.outputs.items()):
+                datasetDataId = commonDataId.subset(datasetType.dimensions)
+                ref = refs.get(datasetDataId)
```
I figured the cost of an extra hash lookup was probably lower than the cost of constructing the temporary if it's usually not needed (as I _think_ it is the case here).  No quantitative analysis of that, though.  Mostly I really wish Python (and every language I've ever used) had a better interface for conditional mapping insertions that avoided that choice.

TallJimbo commented on python/lsst/pipe/base/graphBuilder.py:
```diff
@@ -498,44 +503,61 @@ def fillDataIds(self, registry, collections, userQuery):
+                quantum = task.quanta.get(quantumDataId)
+                if quantum is None:
+                    quantum = _QuantumScaffolding(task=task, dataId=quantumDataId)
+                    task.quanta[quantumDataId] = quantum
```
Same reasoning as above.

TallJimbo commented:
I've addressed or responded to review comments and rebased, and am kicking off a new Jenkins run now since it's been a while.  Will merge when that's green unless I see more comments first.
